src/platform/backends/shared/base_availability_zone_manager.cpp

```diff
@@ -0,0 +1,169 @@
+        {
+            mpl::debug(category, "failed to read AZ manager file': {}", e.what());
+            return QJsonObject{};
+        }
+    };
+
+    const auto json = read_json();
+
+    const auto deserialize_automatic_zone = [&] {
+        const auto json_automatic_zone = json[automatic_zone_key].toString().toStdString();
+        if (std::find(default_zone_names.begin(), default_zone_names.end(), json_automatic_zone) !=
+            default_zone_names.end())
+            return json_automatic_zone;
+
+        mpl::debug(category, "automatic zone '{}' not known, using default", json_automatic_zone);
+        return std::string{*default_zone_names.begin()};
+    };
+
+    return {
+        // TODO remove these comments in C++20
+        /*.file_path = */ file_path,
+        /*.zone_collection = */ {create_default_zones(zones_directory), deserialize_automatic_zone()},
+    };
+}
+
+BaseAvailabilityZoneManager::BaseAvailabilityZoneManager(const fs::path& data_dir)
+    : m{read_from_file(data_dir / az_file, data_dir / zones_directory_name)}
+{
+    serialize();
+}
+
+AvailabilityZone& BaseAvailabilityZoneManager::get_zone(const std::string& name)
+{
+    for (const auto& zone : m.zone_collection.zones)
+    {
+        if (zone->get_name() == name)
+            return *zone;
+    }
+    throw AvailabilityZoneNotFound{name};
+}
+
+std::string BaseAvailabilityZoneManager::get_automatic_zone_name()
+{
+    const auto zone_name = m.zone_collection.next_available();
+    serialize();
+    return zone_name;
+}
+
+std::vector<std::reference_wrapper<const AvailabilityZone>> BaseAvailabilityZoneManager::get_zones()
+{
+    std::vector<std::reference_wrapper<const AvailabilityZone>> zone_list;
+    zone_list.reserve(m.zone_collection.zones.size());
+    for (auto& zone : m.zone_collection.zones)
+        zone_list.emplace_back(*zone);
+    return zone_list;
+}
+
+std::string BaseAvailabilityZoneManager::get_default_zone_name() const
+{
+    return m.zone_collection.zones.begin()->get()->get_name();
+}
+
+void BaseAvailabilityZoneManager::serialize() const
+{
+    mpl::trace(category, "writing AZ manager to file '{}'", m.file_path);
+    const std::unique_lock lock{m.mutex};
+
+    const QJsonObject json{
+        {automatic_zone_key, QString::fromStdString(m.zone_collection.last_used())},
+    };
+
+    MP_JSONUTILS.write_json(json, QString::fromStdString(m.file_path.u8string()));
+}
+
+BaseAvailabilityZoneManager::ZoneCollection::ZoneCollection(
+    std::array<AvailabilityZone::UPtr, default_zone_names.size()>&& _zones,
+    std::string last_used)
+    : zones{std::move(_zones)}, automatic_zone{std::find_if(zones.begin(), zones.end(), [&last_used](const auto& zone) {
+          return zone->get_name() == last_used;
+      })}
+{
+}
+
+std::string BaseAvailabilityZoneManager::ZoneCollection::next_available()
+{
+    const auto start = automatic_zone;
+    auto current = start;
+    do
+    {
+        if (current->get()->is_available())
```

Comment:
We could've used good ol' modulo to avoid the pointer bookkeeping:

```C++
    for(auto i = 0;; i = (++i % stuff.size())){
        const auto & current = stuff[i];
        if(current->get->available()){
            // assign to the outer-scope var
            break;
        }
    }
```
Alternatively, you can leverage the empty conditional to do the available check:

```C++
    foo * current = &stuff[0];
    for(auto i = 1; !current->available(); i = (++i % stuff.size()), current = &stuff[i]){
        std::printf("%s\n", current->value.c_str());
    }
```

... but I think it's a little bit overkill for this scenario. Still, it shows what's possible with a for-loop construct besides the typical uses.